src/Laravel/Eloquent/Metadata/ModelMetadata.php

```diff
@@ -131,13 +131,14 @@ public function getVirtualAttributes(Model $model, array $columns): Collection
                 'name' => $name,
                 'type' => null,
                 'increments' => false,
-                'nullable' => null,
+                'nullable' => true,
                 'default' => null,
                 'unique' => null,
                 'fillable' => $model->isFillable($name),
                 'hidden' => $this->attributeIsHidden($name, $model),
                 'appended' => $model->hasAppended($name),
                 'cast' => $cast,
+                'primary' => $name === $model->getKeyName(),
```

Comment:
The `primary` attribute is accessed for every fields here:
https://github.com/api-platform/core/blob/80fc2d2303f35bccdc6789dc72a205781c3dc3d0/src/Laravel/Eloquent/Metadata/Factory/Property/EloquentPropertyNameCollectionMetadataFactory.php#L55-L56

In MongoDB, "id" ends as a "virtual field" because we have an accessor for it: https://github.com/mongodb/laravel-mongodb/blob/5.0.2/src/Eloquent/DocumentModel.php#L78-L94
Maybe the metadata should be merged when a field is defined both in database schema and as a virtual field.

Reply:
You might also want to do this on line 83 in the same class and remove the `isColumnPrimaryKey` method all along, imho your solution is cleaner.